CRM-17556 Dedupe improvements

JKingsnorth opened the pull request:
* [CRM-17556: Improve manual merge screen to deal with types and location fields](https://issues.civicrm.org/jira/browse/CRM-17556)

JKingsnorth commented:
@todo Support 'website' fields in merging
@todo Fix 'add new' option for phones/emails?
@todo Test fully on 4.7

JKingsnorth commented:
Remaining tasks:
* Support website field
* Fully test on 4.7

JKingsnorth commented:
Todo:
* Ensure tests pass
* Refactoring (I'm sure we can tidy up A LOT)
* Fully test on 4.7

JKingsnorth commented:
Ready for QA

colemanw commented on CRM/Dedupe/Merger.php:
```diff
@@ -907,10 +956,19 @@ public static function skipMerge($mainId, $otherId, &$migrationInfo, $mode = 'sa
+      'provider_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_IM', 'provider_id'),
+      'website_type_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_Website', 'website_type_id'),
+      'gender_id' => CRM_Core_PseudoConstant::get('CRM_Contact_DAO_Contact', 'gender_id'),
+    );
```
This list could quickly get out of control - there are hundreds and hundreds of pseudoconstants in CiviCRM. And hard-coded lists like this are something I've been working to get rid of.
Should be pretty easy to eliminate, since the list is mostly tautological (`'location_type_id'` is mapping `get('CRM_Core_DAO_Address', 'location_type_id')`, etc) all you need to know is the entity which the field belongs to and then you can look it up automatically. Some suggestions:

1. If you're looping through entities like "Contact", "Phone" "Im" etc then you can use the api to fetch pseudoconstants like `civicrm_api3('$entity, 'getoptions', array('field' => $fieldName));` and the api will handle mapping entityName to DAOname for you.
2. To avoid any hardcoding (even hardcoding of whether a field has options or not) our schema already knows all about pseudoconstants - so if you lookup the metadata for an entity you can tell if a field has a "pseudoconstant" property.

colemanw commented on css/civicrm.css:
```diff
@@ -3173,6 +3173,18 @@ div.m ul#civicrm-menu,
+.crm-container table tr.merge-row-equal.crm-row-selected,
+.crm-container table tr.merge-row-diff.crm-row-selected {
+  background-color: #fffdb2;
+}
```
I'm trying to avoid putting css that pertains to a single screen into our global civicrm.css file.
We could put this in admin.css, or better yet, start a new css file just for deduping.

colemanw commented on templates/CRM/Contact/Form/Merge.tpl:
```diff
@@ -126,10 +180,12 @@
     {else}
     <tr class="{cycle values="even-row,odd-row"}">
       <td><strong>{ts}Move related...{/ts}</strong></td><td><a href="{$params.other_url}">{$params.title}</a></td><td style='white-space: nowrap'>=={$form.$paramName.html|crmAddClass:"select-row"}==&gt;</td><td><a href="{$params.main_url}">{$params.title}</a>{if $form.operation.$paramName.add.html}&nbsp;{$form.operation.$paramName.add.html}{/if}</td>
+       <td>(migrate)</td>
```
shouldn't these strings be translated?

colemanw commented:
Code review generally looks good. Made a few minor comments. It's too bad the new js has the same problems as the old (using global functions instead of closures, using the deprecated `cj`, putting `onchange` directly into the html markup instead of using the jQuery `on` method) but oh well.

JKingsnorth commented:
Thanks for the review coleman. I did manage to get rid of the deprecated 'cj' reference at least. These improvements turned up a couple of bugs as well so I had to spend some extra time 'getting things working' rather than refactoring (which I would have liked to do more of). I'll address your comments asap. Thanks.

JKingsnorth commented:
@colemanw Thanks for the review.

* I've fixed up the string translations, and caught a couple of others.
* I've removed the pseudoconst definition lists - they're now only called when needed, and in the context of the entity they're being used in. Is that an improvement?
* Although not quite what you had in mind, I've slimmed down the CSS and made it more generic, so that it could stay with the other row related classes in the main CSS file.

Happy to take any further comments =]

JKingsnorth commented:
Test failure does not seem related.

JKingsnorth commented:
Jenkins, could you re-test this please?
